pennylane/devices/qubit_mixed/apply_operation.py

```diff
@@ -672,3 +679,174 @@ def apply_snapshot(
             debugger.snapshots[len(debugger.snapshots)] = snapshot
 
     return state
+
+
+# pylint: disable=unused-argument
+@apply_operation.register
+def apply_density_matrix(
+    op: qml.QubitDensityMatrix,
+    state,
+    is_state_batched: bool = False,
+    debugger=None,
+    **execution_kwargs,
+):
+    """
+    Applies a QubitDensityMatrix operation by initializing or replacing
+    the quantum state with the provided density matrix.
+
+    - If the QubitDensityMatrix covers all wires, we directly return the provided density matrix as the new state.
+    - If only a subset of the wires is covered, we:
+      1. Partial trace out those wires from the current state to get the density matrix of the complement wires.
+      2. Take the tensor product of the complement density matrix and the provided density_matrix.
+      3. Reshape to the correct final shape and return.
+
+    Args:
+        op (qml.QubitDensityMatrix): The QubitDensityMatrix operation.
+        state (array-like): The current quantum state.
+        is_state_batched (bool): Whether the state is batched.
+        debugger: A debugger instance for diagnostics.
+        **execution_kwargs: Additional kwargs.
+
+    Returns:
+        array-like: The updated quantum state.
+
+    Raises:
+        ValueError: If the density matrix is invalid.
+    """
+    density_matrix = op.parameters[0]
+    num_wires = len(op.wires)
+    expected_dim = 2**num_wires
+
+    # Cast density_matrix to the same type and device as state
+    density_matrix = math.cast_like(density_matrix, state)
+
+    # Validate shape
+    if math.shape(density_matrix) != (expected_dim, expected_dim):
+        raise ValueError(
+            f"Density matrix must have shape {(expected_dim, expected_dim)}, "
+            f"but got {math.shape(density_matrix)}."
+        )
+
+    # Validate Hermiticity
+    if not math.allclose(density_matrix, math.conjugate(math.transpose(density_matrix))):
+        raise ValueError("Density matrix must be Hermitian.")
```

Comment:
Sorry for missing this earlier, but this seems like a check that should be performed in the initialization of `QubitDensityMatrix` rather than here.

Reply:
I'll remove all the three validation and move them to `QubitDensityMatrix` in another PR later.

Reply:
https://github.com/PennyLaneAI/pennylane/pull/6819